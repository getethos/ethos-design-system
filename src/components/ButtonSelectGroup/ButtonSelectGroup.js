import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import uuidv4 from 'uuid/v4'
import useErrorMessage from '../../hooks/useErrorMessage.js'
import { OptionButton } from './OptionButton'
import { InputLabel } from '../InputLabel'

import styles from './ButtonGroup.module.scss'

/**
 **/

/**
 * Component renders a group of button that behaves similarly to a radio group
 *
 * @public
 *
 * @param {object} props - Component Props
 * @prop {string} props.name - Name of field. default value is a uuid
 * @prop {string} props.labelCopy - Set's the caption of the group's label
 * @prop {string} [props.initialValue] - Optionally sets a default value for the group. If set, the matching option will be set as `isSelected`
 * @prop {string} [props.buttonStyle] - Optional value that sets the background color of all the buttons in the group (unselected state)
 * @prop {function} [props.onSelect] - Optional callback thats fires when an option is selected. returns an object containing the selected `value` and a boolean value `isAnswered`
 * @prop {function} [props.formChangeHandler] - Optional callback thats fires when an option is selected. Works similarly to onSelect, but used in `<Form>`.
 *
 * @example ```
 * <ButtonSelectGroup
 *  initialValue="excellent"
 *  labelCopy="Health"
 *  onSelect={({ value }) => console.log(value)}
 * >
 *   <ButtonSelectGroup.Option value="average">Average</ButtonSelectGroup.Option>
 *   <ButtonSelectGroup.Option value="great">Greate</ButtonSelectGroup.Option>
 * </ButtonSelectGroup>
 * ```
 *
 * @return {JSX.Element}
 */
export const ButtonSelectGroup = ({
  labelCopy,
  children,
  initialValue = undefined,
  currentValue,
  currentError,
  formTouched,
  onSelect,
  column,
  formChangeHandler,
  name = `button-select-group-${uuidv4()}`,
  allCaps = true,
  buttonStyle = 'default',
  validator,
  fullWidth = true,
  ...rest
}) => {
  let initialSelected
  if (currentValue || typeof currentValue === 'boolean') {
    initialSelected = currentValue
  } else if (initialValue || typeof initialValue === 'boolean') {
    initialSelected = initialValue
  }

  const [selectedValue, setSelectedValue] = useState(initialSelected)
  const [isAnswered, setIsAnswered] = useState(false)
  // Set up validation hooks
  const [getError, setError, , validate] = useErrorMessage(validator)

  useEffect(() => {
    // `isSelectedValue` allows `false` to work properly and validate
    // For example, if we have two buttons yes & no mapped to booleans
    const isSelectedValue = typeof selectedValue !== 'undefined'
    if (onSelect && isSelectedValue) {
      onSelect({ value: selectedValue, isAnswered })
    }
    if (formChangeHandler && isSelectedValue) {
      // Ensure all validators get called
      let errorMessage = validate(selectedValue)
      errorMessage = errorMessage.length ? errorMessage : ''
      setError(errorMessage)

      // Update form with the new value and a falsy error message
      formChangeHandler(selectedValue, errorMessage)
    }
  }, [selectedValue, isAnswered])

  /**
   * Handles a click event on an descending `<OptionButton/>`
   *
   * @private
   *
   * @param {string} value - the new selected value
   * @param {function} clickHandler - an optional click handler that fires are the value has been set
   *
   * @return {function}
   */
  function onClickHandler(value, clickHandler) {
    return (evt) => {
      // set the selected value to the passed value
      setSelectedValue(value)

      // if the group was not previously answered â€“ set it to true
      if (!isAnswered) {
        setIsAnswered(true)
      }

      return clickHandler && clickHandler(evt)
    }
  }

  // Iterate through each child, inject stateful props
  const options = React.Children.map(children, (child) => {
    const { value, onClick: passedHandler } = child.props

    const isSelected = value === selectedValue
    const onClick = onClickHandler(value, passedHandler)

    return React.cloneElement(child, {
      ...child.props,
      isSelected,
      onClick,
      buttonStyle,
    })
  })

  const optionsContainerClassNames = [styles.buttonGroup]
  if (fullWidth) optionsContainerClassNames.push(styles.fullWidth)
  if (column) optionsContainerClassNames.push(styles.column)

  // Use id to connect label and this pseudo-input because of aria-labelledby
  return (
    <>
      <div
        role="radiogroup"
        aria-labelledby={name}
        className="button-select-group"
        data-tid={rest['data-tid']}
      >
        <InputLabel
          element="span"
          id={name}
          labelCopy={labelCopy}
          allCaps={allCaps}
        />
        <div className={optionsContainerClassNames.join(' ')}>{options}</div>
      </div>
      {getError(currentError, formTouched)}
    </>
  )
}

ButtonSelectGroup.propTypes = {
  formTouched: PropTypes.bool,
  currentValue: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  currentError: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.object, PropTypes.array]).isRequired,
  /** Set's the caption of the group's label */
  labelCopy: PropTypes.string.isRequired,
  /** Name of the field, provided a uuid if not supplied. */
  name: PropTypes.string,
  /** When set to `true`, the group's label will be displayed uppercase */
  allCaps: PropTypes.bool,
  /** Optionally sets a default value for the group. If set, the matching option will be set as `isSelected` */
  initialValue: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  /** Optional value that sets the background color of all the buttons in the group (unselected state) */
  buttonStyle: PropTypes.string,
  /** Optional callback thats fires when an option is selected. returns an object containing the selected `value` and a boolean value `isAnswered` */
  formChangeHandler: PropTypes.func,
  /** Optional callback thats fires when an option is selected. Works similarly to onSelect, but used in `<Form>`. */
  onSelect: PropTypes.func,
  /** When set to `true`, the group will display as flex column */
  column: PropTypes.bool,
  /** Optional data-tid used as a unique id for targeting test selectors */
  'data-tid': PropTypes.string,
  validator: PropTypes.func,
  /** Optional, makes the group width 100%. Defaults to true */
  fullWidth: PropTypes.bool,
}

// Export the option button
ButtonSelectGroup.Option = OptionButton
