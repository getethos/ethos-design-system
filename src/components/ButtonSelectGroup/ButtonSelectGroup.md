### Description
`<ButtonSelectGroup />` renders a group of button that behaves similarly to a radio group


### Examples

### Base Usage

```jsx
<ButtonSelectGroup label="Health">
  <ButtonSelectGroup.Option value="average">Average</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="great">Great</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="excellent">
    Exellent
  </ButtonSelectGroup.Option>
</ButtonSelectGroup>
```

### Rendering with a `defaultValue` selected

```jsx
<ButtonSelectGroup defaultValue="great" label="Health">
  <ButtonSelectGroup.Option value="average">Average</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="great">Great</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="excellent">
    Exellent
  </ButtonSelectGroup.Option>
</ButtonSelectGroup>
```

### Handling `onSelect`

```jsx
<ButtonSelectGroup
  defaultValue="excellent"
  label="Health"
  onSelect={({ value }) => console.log(value)}
>
  <ButtonSelectGroup.Option value="average">Average</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="great">Great</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="excellent">
    Exellent
  </ButtonSelectGroup.Option>
</ButtonSelectGroup>
```


### Handling an `onClick` event on a specific `<Option/>` 

```jsx
<ButtonSelectGroup
  defaultValue="excellent"
  label="Health"
  onSelect={({ value }) => console.log(value)}
>
  <ButtonSelectGroup.Option onClick={() => console.log('i was clicked!')} value="average">Average</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="great">Great</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="excellent">
    Exellent
  </ButtonSelectGroup.Option>
</ButtonSelectGroup>
```