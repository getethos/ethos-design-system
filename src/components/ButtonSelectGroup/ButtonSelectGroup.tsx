import * as React from 'react'
import { useEffect, useState } from 'react'
import { v4 as uuidv4 } from 'uuid'
import useErrorMessage from '../../hooks/useErrorMessage'
import { PublicFormProps } from '../Form/Form'
import { InputLabel } from '../InputLabel'
import styles from './ButtonGroup.module.scss'
import { OptionButton, OptionButtonProps } from './OptionButton'
type ButtonSelectGroupProps = PublicFormProps & {
  currentValue?: string | boolean
  currentError?: string
  labelCopy?: string
  allCaps?: boolean
  capitalize?: boolean
  initialValue?: string | boolean
  buttonStyle?: string
  onSelect?: (...args: any[]) => any
  column?: boolean
  fullWidth?: boolean
}
/**
 **/
/**
 * Component renders a group of button that behaves similarly to a radio group
 *
 * @public
 *
 * @param {object} props - Component Props
 * @prop {string} props.name - Name of field. default value is a uuid
 * @prop {string} props.labelCopy - Set's the caption of the group's label
 * @prop {string} [props.initialValue] - Optionally sets a default value for the group. If set, the matching option will be set as `isSelected`
 * @prop {string} [props.buttonStyle] - Optional value that sets the background color of all the buttons in the group (unselected state)
 * @prop {function} [props.onSelect] - Optional callback thats fires when an option is selected. returns an object containing the selected `value` and a boolean value `isAnswered`
 * @prop {function} [props.formChangeHandler] - Optional callback thats fires when an option is selected. Works similarly to onSelect, but used in `<Form>`.
 *
 * @example ```
 * <ButtonSelectGroup
 *  initialValue="excellent"
 *  labelCopy="Health"
 *  onSelect={({ value }) => console.log(value)}
 * >
 *   <ButtonSelectGroup.Option value="average">Average</ButtonSelectGroup.Option>
 *   <ButtonSelectGroup.Option value="great">Greate</ButtonSelectGroup.Option>
 * </ButtonSelectGroup>
 * ```
 *
 * @return {JSX.Element}
 */
export const ButtonSelectGroup: React.FC<ButtonSelectGroupProps> & {
  Option: typeof OptionButton
} = ({
  labelCopy,
  children,
  initialValue = undefined,
  currentValue,
  currentError,
  formTouched,
  onSelect,
  column,
  formChangeHandler,
  name = `button-select-group-${uuidv4()}`,
  allCaps = true,
  capitalize,
  buttonStyle = 'default',
  validator,
  fullWidth = true,
  ...rest
}) => {
  let initialSelected
  if (currentValue || typeof currentValue === 'boolean') {
    initialSelected = currentValue
  } else if (initialValue || typeof initialValue === 'boolean') {
    initialSelected = initialValue
  }
  const [selectedValue, setSelectedValue] = useState(initialSelected)
  const [isAnswered, setIsAnswered] = useState(false)
  // Set up validation hooks
  const [getError, setError, , validate] = useErrorMessage(validator)
  useEffect(() => {
    // `isSelectedValue` allows `false` to work properly and validate
    // For example, if we have two buttons yes & no mapped to booleans
    const isSelectedValue = typeof selectedValue !== 'undefined'
    if (onSelect && isSelectedValue) {
      onSelect({ value: selectedValue, isAnswered })
    }
    if (formChangeHandler && isSelectedValue) {
      // Ensure all validators get called
      let errorMessage = validate(selectedValue)
      errorMessage =
        (typeof errorMessage === 'string' || Array.isArray(errorMessage)) &&
        errorMessage.length
          ? errorMessage
          : ''
      setError(errorMessage)
      // Update form with the new value and a falsy error message
      formChangeHandler(selectedValue, errorMessage)
    }
  }, [selectedValue, isAnswered])
  /**
   * Handles a click event on an descending `<OptionButton/>`
   *
   * @private
   *
   * @param {string} value - the new selected value
   * @param {function} clickHandler - an optional click handler that fires are the value has been set
   *
   * @return {function}
   */
  function onClickHandler(value, clickHandler) {
    return (evt) => {
      // set the selected value to the passed value
      setSelectedValue(value)
      // if the group was not previously answered â€“ set it to true
      if (!isAnswered) {
        setIsAnswered(true)
      }
      return clickHandler && clickHandler(evt)
    }
  }
  // Iterate through each child, inject stateful props
  const options = React.Children.map(
    children,
    (child: React.ReactElement<OptionButtonProps>) => {
      const { value, onClick: passedHandler } = child.props
      const isSelected = value === selectedValue
      const onClick = onClickHandler(value, passedHandler)

      return React.cloneElement(child, {
        ...child.props,
        isSelected,
        onClick,
        buttonStyle,
      })
    }
  )
  const optionsContainerClassNames = [styles.buttonGroup]
  if (fullWidth) optionsContainerClassNames.push(styles.fullWidth)
  if (column) optionsContainerClassNames.push(styles.column)
  // Use id to connect label and this pseudo-input because of aria-labelledby
  return (
    <>
      <div
        role="radiogroup"
        aria-labelledby={name}
        className="button-select-group"
        data-tid={rest['data-tid']}
      >
        {labelCopy && (
          <InputLabel
            element="span"
            id={name}
            labelCopy={labelCopy}
            allCaps={allCaps}
            capitalize={capitalize}
          />
        )}
        <div className={optionsContainerClassNames.join(' ')}>{options}</div>
      </div>
      {getError(currentError, formTouched)}
    </>
  )
}
// Export the option button
ButtonSelectGroup.Option = OptionButton
