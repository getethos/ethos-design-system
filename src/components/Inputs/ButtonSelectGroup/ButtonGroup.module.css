.button-group {
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;

  & > .Button.Stateful {
    width: 100%;
    margin-right: var(--Space-8);

    &:last-child {
      margin-right: 0;
    }
  }
}
