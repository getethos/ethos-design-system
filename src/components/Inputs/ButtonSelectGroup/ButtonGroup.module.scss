// Button.Stateful is no longer available w/CSS Modules but
// it's always such a button if you look at component JSX
.buttonGroup > button {
  margin-right: var(--Space-8);
}

// fullWidth is default.
.buttonGroup.fullWidth {
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  width: 100%;
  & > button {
    width: 100%;
  }
}

.buttonGroup > button:last-child {
  margin-right: 0;
}
