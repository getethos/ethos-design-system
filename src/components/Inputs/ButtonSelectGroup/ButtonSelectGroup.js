import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import uuidv4 from 'uuid/v4'
import useErrorMessage from '../../../hooks/useErrorMessage.js'
import { OptionButton } from './OptionButton'
import { InputLabel } from '../InputLabel'

import styles from './ButtonGroup.module.scss'
/**
 **/

/**
 * Component renders a group of button that behaves similarly to a radio group
 *
 * @public
 *
 * @param {object} props - Component Props
 * @prop {string} props.name - Name of field. default value is a uuid
 * @prop {string} props.labelCopy - Set's the caption of the group's label
 * @prop {string} [props.defaultValue] - Optionally sets a default value for the group. If set, the matching option will be set as `isSelected`
 * @prop {string} [props.buttonStyle] - Optional value that sets the background color of all the buttons in the group (unselected state)
 * @prop {function} [props.onSelect] - Optional callback thats fires when an option is selected. returns an object containing the selected `value` and a boolean value `isAnswered`
 * @prop {function} [props.formChangeHandler] - Optional callback thats fires when an option is selected. Works similarly to onSelect, but used in `<Form>`.
 *
 * @example ```
 * <ButtonSelectGroup
 *  defaultValue="excellent"
 *  labelCopy="Health"
 *  onSelect={({ value }) => console.log(value)}
 * >
 *   <ButtonSelectGroup.Option value="average">Average</ButtonSelectGroup.Option>
 *   <ButtonSelectGroup.Option value="great">Greate</ButtonSelectGroup.Option>
 * </ButtonSelectGroup>
 * ```
 *
 * @return {JSX.Element}
 */
export const ButtonSelectGroup = ({
  labelCopy,
  children,
  defaultValue,
  onSelect,
  formChangeHandler,
  name = `button-select-group-${uuidv4()}`,
  allCaps = false,
  buttonStyle = 'default',
  validator,
  ...rest
}) => {
  const [selectedValue, setSelectedValue] = useState(defaultValue)
  const [isAnswered, setIsAnswered] = useState(false)
  // Set up validation hooks
  const [getError, setError, validate] = useErrorMessage(validator)

  useEffect(() => {
    if (onSelect && selectedValue) {
      onSelect({ value: selectedValue, isAnswered })
    }
    if (formChangeHandler && selectedValue) {
      // Ensure all validators get called
      let errorMessage = validate(selectedValue)
      errorMessage = errorMessage.length ? errorMessage : ''

      // Update form with the new value and a falsey error message
      formChangeHandler(selectedValue, errorMessage)
    }
  }, [selectedValue])

  /**
   * Handles a click event on an descending `<OptionButton/>`
   *
   * @private
   *
   * @param {string} value - the new selected value
   * @param {function} clickHandler - an optional click handler that fires are the value has been set
   *
   * @return {function}
   */
  function onClickHandler(value, clickHandler) {
    return (evt) => {
      // set the selected value to the passed value
      setSelectedValue(value)

      // if the group was not previously answered â€“ set it to true
      if (!isAnswered) {
        setIsAnswered(true)
      }

      return clickHandler && clickHandler(evt)
    }
  }

  // Iterate through each child, inject stateful props
  const options = React.Children.map(children, (child) => {
    const { value, onClick: passedHandler } = child.props

    const isSelected = value === selectedValue
    const onClick = onClickHandler(value, passedHandler)

    return React.cloneElement(child, {
      ...child.props,
      isSelected,
      onClick,
      buttonStyle,
    })
  })

  // Use id to connect label and this pseudo-input because of aria-labelledby
  return (
    <div
      role="radiogroup"
      aria-labelledby={name}
      className={styles['button-select-group']}
      data-tid={rest['data-tid']}
    >
      <InputLabel
        element="span"
        id={name}
        labelCopy={labelCopy}
        allCaps={allCaps}
      />
      <div className={styles['button-group']}>{options}</div>
    </div>
  )
}

ButtonSelectGroup.propTypes = {
  /** Set's the caption of the group's label */
  labelCopy: PropTypes.string.isRequired,
  /** Name of the field, provided a uuid if not supplied. */
  name: PropTypes.string,
  /** When set to `true`, the group's label will be displayed uppercase */
  allCaps: PropTypes.bool,
  /** Optionally sets a default value for the group. If set, the matching option will be set as `isSelected` */
  defaultValue: PropTypes.string,
  /** Optional value that sets the background color of all the buttons in the group (unselected state) */
  buttonStyle: PropTypes.string,
  /** Optional callback thats fires when an option is selected. returns an object containing the selected `value` and a boolean value `isAnswered` */
  formChangeHandler: PropTypes.func,
  /** Optional callback thats fires when an option is selected. Works similarly to onSelect, but used in `<Form>`. */
  onSelect: PropTypes.func,
  /** Optional data-tid used as a unique id for targeting test selectors */
  'data-tid': PropTypes.string,
  validator: PropTypes.func,
}

// Export the option button
ButtonSelectGroup.Option = OptionButton
