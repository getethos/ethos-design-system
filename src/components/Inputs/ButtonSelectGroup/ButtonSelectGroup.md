### Description

`<ButtonSelectGroup />` renders a group of button that behaves similarly to a radio group

### Examples

### Base Usage

```jsx
<ButtonSelectGroup label="Health">
  <ButtonSelectGroup.Option value="average">Average</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="great">Great</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="excellent">
    Exellent
  </ButtonSelectGroup.Option>
</ButtonSelectGroup>
```

### Rendering with a `defaultValue` selected

```jsx
<ButtonSelectGroup defaultValue="great" label="Health">
  <ButtonSelectGroup.Option value="average">Average</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="great">Great</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="excellent">
    Exellent
  </ButtonSelectGroup.Option>
</ButtonSelectGroup>
```

### Handling `onSelect`

```jsx
<ButtonSelectGroup
  defaultValue="excellent"
  labelCopy="Health"
  onSelect={({ value }) => console.log(value)}
>
  <ButtonSelectGroup.Option value="average">Average</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="great">Great</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="excellent">
    Exellent
  </ButtonSelectGroup.Option>
</ButtonSelectGroup>
```

### Handling an `onClick` event on a specific `<Option/>`

```jsx
<ButtonSelectGroup
  defaultValue="excellent"
  labelCopy="Health"
  onSelect={({ value }) => console.log(value)}
>
  <ButtonSelectGroup.Option
    onClick={() => console.log('i was clicked!')}
    value="average"
  >
    Average
  </ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="great">Great</ButtonSelectGroup.Option>
  <ButtonSelectGroup.Option value="excellent">
    Exellent
  </ButtonSelectGroup.Option>
</ButtonSelectGroup>
```

### Styling `<OptionButton />`

`<OptionButtons />` can be styled by applying the `buttonStyle` property to the wrapping `<ButtonSelectGroup />`. The `buttonStyle` property expects to be passed a valid style enumeration option from the `OPTION_BUTTON_STYLES` `ENUM` (imported from the `BUTTON_SELECT_GROUP` package).

There are currently two options: `WHITE` and `DEFAULT`. If `buttonStyle` is not passed, the button style is defaulted to `DEFAULT`.

### White `buttonStyle`

```jsx
import { OPTION_BUTTON_STYLES } from './index.js'
;<div style={{ background: '#f5f5f5', padding: '8px 16px' }}>
  <ButtonSelectGroup
    defaultValue="riddikulus"
    labelCopy="Cast Spell"
    buttonStyle={OPTION_BUTTON_STYLES.WHITE}
  >
    <ButtonSelectGroup.Option value="riddikulus">
      Riddikulus
    </ButtonSelectGroup.Option>
    <ButtonSelectGroup.Option value=" expecto_patronum">
      Expecto Patronum
    </ButtonSelectGroup.Option>
  </ButtonSelectGroup>
</div>
```

### Default `buttonStyle`

```jsx
import { OPTION_BUTTON_STYLES } from './index.js'
;<div style={{ background: '#f5f5f5', padding: '8px 16px' }}>
  <ButtonSelectGroup
    defaultValue="riddikulus"
    labelCopy="Cast Spell"
    buttonStyle={OPTION_BUTTON_STYLES.DEFAULT}
  >
    <ButtonSelectGroup.Option value="riddikulus">
      Riddikulus
    </ButtonSelectGroup.Option>
    <ButtonSelectGroup.Option value=" expecto_patronum">
      Expecto Patronum
    </ButtonSelectGroup.Option>
  </ButtonSelectGroup>
</div>
```
