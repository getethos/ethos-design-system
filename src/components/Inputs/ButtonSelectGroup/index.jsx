export { OPTION_BUTTON_STYLES } from './OptionButton'

export { ButtonSelectGroup } from './ButtonSelectGroup'
