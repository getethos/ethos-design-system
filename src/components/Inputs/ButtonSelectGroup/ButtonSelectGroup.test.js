import React from 'react'
import { OPTION_BUTTON_STYLES } from './OptionButton'
import { ButtonSelectGroup } from './ButtonSelectGroup'
import renderer from 'react-test-renderer'

jest.mock('uuid/v4', () => {
  return jest.fn(() => 1)
})

describe('ButtonSelectGroup', () => {
  test('The ButtonSelectGroup renders (default button style, `buttonStyle` not passed)', () => {
    const tree = renderer
      .create(
        <ButtonSelectGroup labelCopy="options">
          <ButtonSelectGroup.Option value="foo">foo</ButtonSelectGroup.Option>
          <ButtonSelectGroup.Option value="bar">bar</ButtonSelectGroup.Option>
        </ButtonSelectGroup>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('The ButtonSelectGroup renders (default button style, `buttonStyle` passed)', () => {
    const tree = renderer
      .create(
        <ButtonSelectGroup
          buttonStyle={OPTION_BUTTON_STYLES.DEFAULT}
          labelCopy="options"
        >
          <ButtonSelectGroup.Option value="foo">foo</ButtonSelectGroup.Option>
          <ButtonSelectGroup.Option value="bar">bar</ButtonSelectGroup.Option>
        </ButtonSelectGroup>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('The ButtonSelectGroup renders (white button style)', () => {
    const tree = renderer
      .create(
        <ButtonSelectGroup
          buttonStyle={OPTION_BUTTON_STYLES.WHITE}
          labelCopy="options"
        >
          <ButtonSelectGroup.Option value="foo">foo</ButtonSelectGroup.Option>
          <ButtonSelectGroup.Option value="bar">bar</ButtonSelectGroup.Option>
        </ButtonSelectGroup>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('When passed an `onSelect` handler, the callback fires when an Option is `clicked`', () => {
    const onSelectStub = jest.fn()
    const optionValues = ['foo', 'bar']

    const optionButtons = optionValues.map((v, i) => {
      return (
        <ButtonSelectGroup.Option key={i} value={v}>
          {v}
        </ButtonSelectGroup.Option>
      )
    })
    const tree = renderer.create(
      <ButtonSelectGroup labelCopy="placeholder" onSelect={onSelectStub}>
        {optionButtons}
      </ButtonSelectGroup>
    )

    // Grab the group
    const group = tree.root
    const options = group.findAllByType(ButtonSelectGroup.Option)

    const [first, ...rest] = options

    renderer.act(() => {
      first.props.onClick()
    })

    expect(onSelectStub).toBeCalledWith({
      value: optionValues[0],
      isAnswered: true,
    })
  })

  test('When passed a `defaultValue` renders with that value selected, and sets a new option as selected when clicked', () => {
    const onSelectStub = jest.fn()
    const optionValues = ['foo', 'bar']

    const optionButtons = optionValues.map((v, i) => {
      return (
        <ButtonSelectGroup.Option key={i} value={v}>
          {v}
        </ButtonSelectGroup.Option>
      )
    })

    const tree = renderer.create(
      <ButtonSelectGroup
        labelCopy="options"
        defaultValue={optionValues[0]}
        onSelect={onSelectStub}
      >
        {optionButtons}
      </ButtonSelectGroup>
    )

    // Grab the group
    const group = tree.root
    const options = group.findAllByType(ButtonSelectGroup.Option)

    const [first, last] = options

    expect(first.props.isSelected).toBe(true)

    renderer.act(() => {
      last.props.onClick()
    })

    expect(onSelectStub).toBeCalledWith({
      value: optionValues[1],
      isAnswered: true,
    })
  })

  test('When a click handler is passed to an option, it fires', () => {
    const onSelectStub = jest.fn()
    const onClickStub = jest.fn()
    const value = 'foo'

    const tree = renderer.create(
      <ButtonSelectGroup labelCopy="options" onSelect={onSelectStub}>
        <ButtonSelectGroup.Option value={value} onClick={onClickStub}>
          {value}
        </ButtonSelectGroup.Option>
      </ButtonSelectGroup>
    )

    // Grab the group
    const group = tree.root
    const option = group.findByType(ButtonSelectGroup.Option)

    renderer.act(() => {
      option.props.onClick()
    })

    expect(onClickStub).toBeCalled()
    expect(onSelectStub).toBeCalledWith({
      value,
      isAnswered: true,
    })
  })
})
