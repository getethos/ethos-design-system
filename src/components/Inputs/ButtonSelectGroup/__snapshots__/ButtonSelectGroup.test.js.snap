// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonSelectGroup The ButtonSelectGroup renders (default button style, \`buttonStyle\` not passed) 1`] = `
Array [
  <div
    aria-labelledby="button-select-group-1"
    className="button-select-group"
    role="radiogroup"
  >
    <span
      className="Caption Theinhardt Medium500 GrayPrimary AllCaps"
      id="button-select-group-1"
    >
      options
    </span>
    <div
      aria-hidden={true}
      className="Spacer"
      style={
        Object {
          "height": 8,
        }
      }
    />
    <div
      className="buttonGroup fullWidth"
    >
      <button
        className="Button Medium Stateful"
        onClick={[Function]}
        type="button"
      >
        <span
          id="selection-option-1"
        >
          foo
        </span>
      </button>
      <button
        className="Button Medium Stateful"
        onClick={[Function]}
        type="button"
      >
        <span
          id="selection-option-1"
        >
          bar
        </span>
      </button>
    </div>
  </div>,
  "",
]
`;

exports[`ButtonSelectGroup The ButtonSelectGroup renders (default button style, \`buttonStyle\` passed) 1`] = `
Array [
  <div
    aria-labelledby="button-select-group-1"
    className="button-select-group"
    role="radiogroup"
  >
    <span
      className="Caption Theinhardt Medium500 GrayPrimary AllCaps"
      id="button-select-group-1"
    >
      options
    </span>
    <div
      aria-hidden={true}
      className="Spacer"
      style={
        Object {
          "height": 8,
        }
      }
    />
    <div
      className="buttonGroup fullWidth"
    >
      <button
        className="Button Medium Stateful"
        onClick={[Function]}
        type="button"
      >
        <span
          id="selection-option-1"
        >
          foo
        </span>
      </button>
      <button
        className="Button Medium Stateful"
        onClick={[Function]}
        type="button"
      >
        <span
          id="selection-option-1"
        >
          bar
        </span>
      </button>
    </div>
  </div>,
  "",
]
`;

exports[`ButtonSelectGroup The ButtonSelectGroup renders (white button style) 1`] = `
Array [
  <div
    aria-labelledby="button-select-group-1"
    className="button-select-group"
    role="radiogroup"
  >
    <span
      className="Caption Theinhardt Medium500 GrayPrimary AllCaps"
      id="button-select-group-1"
    >
      options
    </span>
    <div
      aria-hidden={true}
      className="Spacer"
      style={
        Object {
          "height": 8,
        }
      }
    />
    <div
      className="buttonGroup fullWidth"
    >
      <button
        className="Button Medium Stateful White"
        onClick={[Function]}
        type="button"
      >
        <span
          id="selection-option-1"
        >
          foo
        </span>
      </button>
      <button
        className="Button Medium Stateful White"
        onClick={[Function]}
        type="button"
      >
        <span
          id="selection-option-1"
        >
          bar
        </span>
      </button>
    </div>
  </div>,
  "",
]
`;
